### PR TITLE
Feat: implement lst()

### LINKED ISSUE #93: incorporate `tibble::lst` ?
"Self-naming lists" are an amazingly handy and surprisingly missing component of base R. This is available in `Hmisc::llist()` and in a non-exported function in `lme4` (`lme4:::namedList`) (see also [this StackOverflow question](https://stackoverflow.com/questions/16951080/can-lists-be-created-that-name-themselves-based-on-input-object-names/16951524#16951524 )). `tibble::lst()` also does sequential construction (i.e. later values on the list can depend on values of elements earlier in the list), which seems handy and worth implementing for compatibility with `tibble::lst` ...

If you want a PR let me know and I will see what I can do (the naming parts are easy, I haven't tried the sequential construction stuff but seems straightforward enough?)

### PR DESCRIPTION
**Pull request by etiennebacher**
Close #93. This PR builds on the code that you have written in #93, I just added a few things to deal with names and to match with the examples of `tibble::lst()`. For the tests, I only took those from `tibble::lst()`.

Examples:
``` r
suppressPackageStartupMessages({
  library(poorman)
})

lst(n = 5, x = runif(n))
#> $n
#> [1] 5
#> 
#> $x
#> [1] 0.8556188 0.4592575 0.8778909 0.6046387 0.5285325

lst(1:3, z = letters[4:6], runif(3))
#> $`1:3`
#> [1] 1 2 3
#> 
#> $z
#> [1] "d" "e" "f"
#> 
#> $`runif(3)`
#> [1] 0.9535917 0.5555817 0.6382548

a <- 1:3
b <- letters[4:6]
lst(a, b)
#> $a
#> [1] 1 2 3
#> 
#> $b
#> [1] "d" "e" "f"

lst(b = 1, a = b, a = b + 1, b = a)
#> $b
#> [1] 1
#> 
#> $a
#> [1] 1
#> 
#> $a
#> [1] 2
#> 
#> $b
#> [1] 2
```

<sup>Created on 2022-08-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

### CODE REVIEW

File: R/lst.R [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+drop_dup_list <- function(x) {
+
+  list_names <- names(x)
+  if (identical(list_names, unique(list_names))) return(x)
+
+  count <- table(list_names)
+  dupes <- names(count[count > 1])
+  uniques <- names(count[count == 1])
+
+  to_drop <- c()
```
Comment (nathaneastwood) [inline R/lst.R:83]:
```suggestion
  to_drop <- vector(NA, length = length(dupes))
```

maybe? Or you could do the whole thing with an apply function

```r
  to_drop <-   do.call(c, lapply(
    dupes,
    function(x) {
      matches <- which(list_names == i)
      matches[-length(matches)]
    }
  ))
```

You could potentially make the internals a little simpler too with

```r
  which.max(list_names == "a")
  # [1] 1
```

File: R/lst.R [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  for (i in dupes) {
+    matches <- which(list_names == i)
+    to_drop <- c(to_drop, matches[-length(matches)])
+  }
+  for (i in uniques) {
+    if (is.null(x[[i]])) {
+      x[[i]] <- NULL
+    }
+  }
```
Comment (nathaneastwood) [inline R/lst.R:92]: I'm not completely sure what this is for however you could maybe look at

```suggestion
  x[uniques] <- Filter(Negate(is.null), x[uniques])
```

**Review by nathaneastwood — CHANGES_REQUESTED** [review]
Hi @etiennebacher. Thanks again for your great work on this. I have a couple of comments where I think things could be simplified a little. Otherwise it looks great.

File: R/lst.R (hunk 1 above)
Comment (etiennebacher) [reply R/lst.R:83]: Right, I'll use your `do.call` code. However, for the internals, strangely `which.max` doesn't work as expected:
``` r
list_names <- c("a", "b", "a")
which.max(list_names == "a")
#> [1] 1
max(which(list_names == "a"))
#> [1] 3
```

File: R/lst.R [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+          listToEval
+        } else {
+          # restrict the environment to the previous elements of the list (and
+          # to the last value for each name if there are duplicated names)
+          drop_dup_list(out[1:(element - 1)])
+        }
+      )
+    }
+    if (is.null(value)) {
+      out[element] <- list(NULL) # set a list element to NULL
```
[the remark this answers is not in the document]
Comment (etiennebacher) [reply R/lst.R:50]: Indeed, I remember that I was surprised for this because I initially tried `out[element] <- NULL`

**Comment by etiennebacher** [comment]
@nathaneastwood thanks! I think I've addressed all your comments

**Comment by nathaneastwood** [comment]
Thanks a lot

**Pull request merged**